src/blenderbim/blenderbim/bim/module/material/operator.py

```diff
@@ -124,21 +124,26 @@ def execute(self, context):
         return IfcStore.execute_ifc_operator(self, context)
 
     def _execute(self, context):
-        obj = bpy.data.objects.get(self.obj) if self.obj else context.active_object
-        material_type = self.material_type or obj.BIMObjectMaterialProperties.material_type
         self.file = IfcStore.get_file()
-        element = self.file.by_id(obj.BIMObjectProperties.ifc_definition_id)
-        ifcopenshell.api.run(
-            "material.assign_material",
-            self.file,
-            **{
-                "product": element,
-                "type": material_type,
-                "material": self.file.by_id(int(obj.BIMObjectMaterialProperties.material)),
-            },
-        )
-        Data.load(IfcStore.get_file())
-        Data.load(IfcStore.get_file(), obj.BIMObjectProperties.ifc_definition_id)
+        active_object = bpy.data.objects.get(self.obj, context.active_object)
+        objs = [bpy.data.objects.get(self.obj)] if self.obj else context.selected_objects
+        material = self.file.by_id(int(active_object.BIMObjectMaterialProperties.material))
+        material_type = self.material_type or active_object.BIMObjectMaterialProperties.material_type
+        for obj in objs:
+            obj_id = obj.BIMObjectProperties.ifc_definition_id
+            if not obj_id:
+                continue
+            ifcopenshell.api.run(
+                "material.assign_material",
+                self.file,
+                **{
+                    "product": self.file.by_id(obj_id),
+                    "type": material_type,
+                    "material": material,
+                },
+            )
+            Data.load(self.file)
+            Data.load(self.file, obj_id)
```

Comment:
Revert - materials, like geometry is complex. Materials can actually be used to parametrically define geometry (e.g. if a material has layers such as for a wall or slab, it auto updates prism-like geometries of objects. I'd rather not make this change and approach it slowly and carefully.